### ISSUE TITLE
Invalid header error when flashing with Secure Boot + Flash Encryption + custom partition table (IDFGH-3869)

### ISSUE THREAD

**Issue by achraf-boussaada** [espressif/esp-idf#5771]
## Environment

- Development Kit:    none
- Module or chip used:  ESP32-WROOM-32E
- IDF version : ESP-IDF v4.3-dev-907-g6c17e3a64
- Build System:         idf.py
- Compiler version : (crosstool-NG esp-2020r2) 8.2.0
- Operating System:    Linux
- Using an IDE?: yes, visual code studio
- Power Supply:        USB

## Problem Description

I managed to activate Flash Encryption with a custom table and now I want to use Secure Boot alongside them.
I'm using the a new chip with the module mentioned above. 
these are the following steps I did on a fresh installation with the hello world project:

**idf.py menuconfig:**

1. Security features --> Enable flash encryption on boot (release mode)
2. Security features --> Enable hardware Secure Boot in bootloader
3. Partition Table --> Change partition table from 0x8000 to 0x9000
4. Partition Table --> Custom partition from table CSV
5. Serial flasher config --> changed Flash size to 4MB
6. Bootloader config --> Bootloader log verbosity (Error)
7. Component config --> NVS --> Enable NVS encryption

**This is my custom partition table:** 
```
# Name,   Type, SubType, Offset,  Size, Flags
# Note: if you have increased the bootloader size, make sure to update the offsets to avoid overlap
nvs,      data, nvs,     ,        0x6000,
nvs_private,data,nvs,,0x3000,
otadata,data,ota,,0x2000,
phy_init, data, phy,     ,        0x1000,
factory,  app,  factory, ,        1M,
ota_0,    app,    ota_0,, 1M,
ota_1,    app,    ota_1,, 1M,
private_keys,data,nvs_keys,,0x1000,encrypted
```
I generated it then using:
```
$IDF_PATH/components/nvs_flash/nvs_partition_generator/nvs_partition_gen.py encrypt nvs_private.csv nvs_private.bin 0x3000 --keygen --keyfile private_keys.bin
```

I generated the private signing key for Secure boot using:
```
espsecure.py generate_signing_key secure_boot_signing_key.pem
```

I flashed the bootloader using the generated command from `idf.py build`:
```
esptool.py --before default_reset --after no_reset write_flash --flash_mode dio --flash_freq 40m --flash_size 4MB 0x1000 /workspace/esp/hello_world/build/bootloader/bootloader.bin
```
I flashed the data of the custom partition table and the keys of the encrypted partition as follows:
```
esptool.py --before default_reset --after no_reset write_flash 0x10000 nvs_private.bin
```
```
esptool.py --before default_reset --after no_reset write_flash 0x320000 keys/private_keys.bin
```

Finally I tried to complete the flash process using `idf.py flash monitor` and I got the following error that keeps showing until I exist manually:
```
rst:0x10 (RTCWDT_RTC_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
invalid header: 0xe3cbd1f7
invalid header: 0xe3cbd1f7
invalid header: 0xe3cbd1f7
invalid header: 0xe3cbd1f7
invalid header: 0xe3cbd1f7
invalid header: 0xe3cbd1f7
invalid header: 0xe3cbd1f7
invalid header: 0xe3cbd1f7

```
I tried to increase the bootloader size by changing the partition table offset and an other time by changing the bootloader verbosity level but I still get the same error but with a different header value.

Is the order of which I flash the bootloader and my custom table matters? or is there a problem with my partition table that needs to be fixed?

**Comment by achraf-boussaada** [comment]
I tried the same steps on a new chip but the same model as mentioned in the issue descritipon and I got a similar problem. This time I can't connect to the chip anymore.

here's the output after running `idf.py flash monitor`:
```
--- idf_monitor on /dev/ttyUSB0 115200 ---
--- Quit: Ctrl+] | Menu: Ctrl+T | Help: Ctrl+T followed by Ctrl+H ---
ets Jul 29 2019 12:21:46

rst:0x1 (POWERON_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:2
load:0x3fff0030,len:20
load:0x3fff0044,len:12020
load:0x40078000,len:20084
ho 0 tail 12 room 4
load:0x40080400,len:3952
0x40080400: _init at ??:?

entry 0x40080678
I (33) boot: ESP-IDF v4.3-dev-907-g6c17e3a64-dirty 2nd stage bootloader
I (34) boot: compile time 11:58:06
I (35) boot: chip revision: 3
I (38) boot.esp32: SPI Speed      : 40MHz
I (43) boot.esp32: SPI Mode       : DIO
I (48) boot.esp32: SPI Flash Size : 4MB
I (52) boot: Enabling RNG early entropy source...
I (58) boot: Partition Table:
I (61) boot: ## Label            Usage          Type ST Offset   Length
I (68) boot:  0 nvs              WiFi data        01 02 0000f000 00006000
I (76) boot:  1 nvs_private      WiFi data        01 02 00015000 00003000
I (83) boot:  2 otadata          OTA data         01 00 00018000 00002000
I (91) boot:  3 phy_init         RF data          01 01 0001a000 00001000
I (98) boot:  4 factory          factory app      00 00 00020000 00100000
I (106) boot:  5 ota_0            OTA app          00 10 00120000 00100000
I (113) boot:  6 ota_1            OTA app          00 11 00220000 00100000
I (121) boot:  7 private_keys     NVS keys         01 04 00320000 00001000
I (128) boot: End of partition table
I (133) boot: Defaulting to factory image
I (137) esp_image: segment 0: paddr=0x00020020 vaddr=0x3f400020 size=0x0594c ( 22860) map
I (155) esp_image: segment 1: paddr=0x00025974 vaddr=0x3ffb0000 size=0x02270 (  8816) load
I (159) esp_image: segment 2: paddr=0x00027bec vaddr=0x40080000 size=0x00404 (  1028) load
0x40080000: _WindowOverflow4 at /workspace/esp/esp-idf/components/freertos/xtensa/xtensa_vectors.S:1730

I (165) esp_image: segment 3: paddr=0x00027ff8 vaddr=0x40080404 size=0x08020 ( 32800) load
I (187) esp_image: segment 4: paddr=0x00030020 vaddr=0x400d0020 size=0x13dd0 ( 81360) map
0x400d0020: _stext at ??:?

I (219) esp_image: segment 5: paddr=0x00043df8 vaddr=0x40088424 size=0x01a54 (  6740) load
0x40088424: rtc_clk_cpu_freq_to_pll_mhz at /workspace/esp/esp-idf/components/soc/src/esp32/rtc_clk.c:495

I (222) esp_image: segment 6: paddr=0x00045854 vaddr=0x00000000 size=0x0a77c ( 42876) 
I (241) esp_image: Verifying image signature...
I (241) secure_boot: Secure Boot eFuse bit(ABS_DONE_1) not yet programmed.
I (243) secure_boot: Verifying with RSA-PSS...
I (258) boot: Loaded app from partition at offset 0x20000
I (258) secure_boot_v2: enabling secure boot v2...
I (259) esp_image: segment 0: paddr=0x00001020 vaddr=0x3fff0030 size=0x00014 (    20) 
I (268) esp_image: segment 1: paddr=0x0000103c vaddr=0x3fff0044 size=0x02ef4 ( 12020) 
I (281) esp_image: segment 2: paddr=0x00003f38 vaddr=0x40078000 size=0x04e74 ( 20084) 
I (293) esp_image: segment 3: paddr=0x00008db4 vaddr=0x40080400 size=0x00f70 (  3952) 
0x40080400: _init at ??:?

I (295) esp_image: Verifying image signature...
I (299) secure_boot: Secure Boot eFuse bit(ABS_DONE_1) not yet programmed.
I (307) secure_boot: Verifying with RSA-PSS...
I (327) secure_boot_v2: valid signature block found
I (332) secure_boot_v2: Burning public key hash to efuse.
I (332) secure_boot_v2: Write protecting public key digest...
I (390) secure_boot_v2: valid signature block found
I (395) secure_boot_v2: blowing secure boot efuse...
I (395) secure_boot_v2: Disable JTAG...
I (395) secure_boot_v2: Disable ROM BASIC interpreter fallback...
I (401) secure_boot_v2: Disable ROM Download mode...
I (430) secure_boot_v2: secure boot v2 is now enabled.
I (430) boot: Checking flash encryption...
I (430) flash_encrypt: Generating new flash encryption key...
I (436) flash_encrypt: Read & write protecting new key...
I (441) flash_encrypt: Setting CRYPT_CONFIG efuse to 0xF
I (447) flash_encrypt: Disable UART bootloader encryption...
I (454) flash_encrypt: Disable UART bootloader decryption...
I (460) flash_encrypt: Disable UART bootloader MMU cache...
I (466) flash_encrypt: Disable JTAG...
I (471) flash_encrypt: Disable ROM BASIC interpreter fallback...
I (489) esp_image: segment 0: paddr=0x00001020 vaddr=0x3fff0030 size=0x00014 (    20) 
I (489) esp_image: segment 1: paddr=0x0000103c vaddr=0x3fff0044 size=0x02ef4 ( 12020) 
I (501) esp_image: segment 2: paddr=0x00003f38 vaddr=0x40078000 size=0x04e74 ( 20084) 
I (513) esp_image: segment 3: paddr=0x00008db4 vaddr=0x40080400 size=0x00f70 (  3952) 
0x40080400: _init at ??:?

I (516) esp_image: Verifying image signature...
I (520) secure_boot: Verifying with RSA-PSS...
Sig block 0 signed with untrusted key
E (528) secure_boot: Secure Boot V2 verification failed.
E (534) esp_image: Secure boot signature verification failed
I (540) esp_image: Calculating simple hash to check for corruption...
W (558) esp_image: image valid, signature bad
W (558) flash_encrypt: no valid bootloader was found
I (611) flash_encrypt: Encrypting partition 2 at offset 0x18000...
I (665) esp_image: segment 0: paddr=0x00020020 vaddr=0x3f400020 size=0x0594c ( 22860) map
I (674) esp_image: segment 1: paddr=0x00025974 vaddr=0x3ffb0000 size=0x02270 (  8816) 
I (677) esp_image: segment 2: paddr=0x00027bec vaddr=0x40080000 size=0x00404 (  1028) 
0x40080000: _WindowOverflow4 at /workspace/esp/esp-idf/components/freertos/xtensa/xtensa_vectors.S:1730

I (682) esp_image: segment 3: paddr=0x00027ff8 vaddr=0x40080404 size=0x08020 ( 32800) 
I (703) esp_image: segment 4: paddr=0x00030020 vaddr=0x400d0020 size=0x13dd0 ( 81360) map
0x400d0020: _stext at ??:?

I (734) esp_image: segment 5: paddr=0x00043df8 vaddr=0x40088424 size=0x01a54 (  6740) 
0x40088424: rtc_clk_cpu_freq_to_pll_mhz at /workspace/esp/esp-idf/components/soc/src/esp32/rtc_clk.c:495

I (737) esp_image: segment 6: paddr=0x00045854 vaddr=0x00000000 size=0x0a77c ( 42876) 
I (757) esp_image: Verifying image signature...
I (757) secure_boot: Verifying with RSA-PSS...
Sig block 0 signed with untrusted key
E (759) secure_boot: Secure Boot V2 verification failed.
E (765) esp_image: Secure boot signature verification failed
I (771) esp_image: Calculating simple hash to check for corruption...
W (834) esp_image: image valid, signature bad
E (835) esp_image: image at 0x120000 has invalid magic byte
W (835) esp_image: image at 0x120000 has invalid SPI mode 255
W (841) esp_image: image at 0x120000 has invalid SPI size 15
E (847) esp_image: image at 0x220000 has invalid magic byte
W (853) esp_image: image at 0x220000 has invalid SPI mode 255
W (860) esp_image: image at 0x220000 has invalid SPI size 15
I (866) flash_encrypt: Encrypting partition 7 at offset 0x320000...
I (927) flash_encrypt: Setting FLASH_CRYPT_CNT for permanent encryption
I (939) flash_encrypt: Flash encryption completed
I (939) boot: Resetting with flash encryption enabled...
ets Jul 29 2019 12:21:46

rst:0x3 (SW_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
invalid header: 0xc4dfbb5a
invalid header: 0xc4dfbb5a
invalid header: 0xc4dfbb5a
invalid header: 0xc4dfbb5a
invalid header: 0xc4dfbb5a
invalid header: 0xc4dfbb5a
invalid header: 0xc4dfbb5a
invalid header: 0xc4dfbb5a
invalid header: 0xc4dfbb5a
invalid header: 0xc4dfbb5a
invalid header: 0xc4dfbb5a
invalid header: 0xc4dfbb5a
invalid header: 0xc4dfbb5a
invalid header: 0xc4dfbb5a
ets Jul 29 2019 12:21:46

rst:0x7 (TG0WDT_SYS_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
invalid header: 0xc4dfbb5a
invalid header: 0xc4dfbb5a
invalid header: 0xc4dfbb5a
invalid header: 0xc4dfbb5a
invalid header: 0xc4dfbb5a
invalid header: 0xc4dfbb5a
invalid header: 0xc4dfbb5a
invalid header: 0xc4dfbb5a
invalid header: 0xc4dfbb5a
invalid header: 0xc4dfbb5a
invalid header: 0xc4dfbb5a
invalid header: 0xc4dfbb5a
invalid header: 0xc4dfbb5a
invalid header: 0xc4dfbb5a
ets Jul 29 2019 12:21:46

rst:0x7 (TG0WDT_SYS_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
invalid header: 0xc4dfbb5a
invalid header: 0xc4dfbb5a
invalid header: 0xc4dfbb5a
invalid header: 0xc4dfbb5a
invalid header: 0xc4dfbb5a
invalid header: 0xc4dfbb5a
invalid header: 0xc4dfbb5a
invalid header: 0xc4dfbb5a
invalid header: 0xc4dfbb5a
invalid header: 0xc4dfbb5a
invalid header: 0xc4dfbb5a
invalid header: 0xc4dfbb5a
invalid header: 0xc4dfbb5a
invalid header: 0xc4dfbb5a
ets Jul 29 2019 12:21:46
```
**NOTE:** I changed the chip on the board I'm using and I was able to establish a connection.

**Comment by projectgus** [comment]
Hi @achraf-boussaada ,

I'm afraid you've encountered a recent regression with Secure Boot V2 on the master branch. By default, Secure Boot V2 also enables the feature to [disable ROM download mode](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/kconfig.html#config-secure-disable-rom-dl-mode) but there is a bug burning this eFuse on first boot and as a result the ESP32 ends up in a bad state. I'm afraid that these ESP32s are not recoverable, as download mode is disabled.

We have a workaround and a follow-up fix for this, we're testing both now and will be released on master branch ASAP. I'm really sorry for the confusion.

Are you looking to go into production? If so, we would generally recommend targeting a stable release version such as [v4.1](https://github.com/espressif/esp-idf/releases/tag/v4.1) (which supports Secure Boot V2 on ESP32-V3) rather than a master branch snapshot. Although typically we don't anticipate bugs this severe even in master branch, this is 100% failure on our part.

Alternatively as a workaround with the version you have it is possible to enable "Security -> Allow potentially insecure options" and then enable "Potentially insecure options -> Don’t automatically restrict UART download mode", double-check that "Permanently disable ROM Download Mode" is not enabled, and then the configuration you have should work on a new chip without any further problems.

I'm very sorry for the inconvenience. This issue will be updated as soon as the bugfix is released on master branch.

**Comment by achraf-boussaada** [comment]
Hi @projectgus ,

I managed to achieve my goal using your workaround. Thanks for the tip. 
We indeed tend to go to production and we're using the master branch since there are some things needed for our project that are not yet available (or we couldn't make it work) with the latest stable version. 

I have now two questions on my mind:

1. Does applying this workaround has any security implications?

2. The BLK2 contains now the Secure Boot key. The problem here is that the key itself is visible when calling `espefuse.py summary` and not replaced by question marks. Is it meant to be like that? Here's the eFuses table after flashing:
```
Security fuses:
FLASH_CRYPT_CNT (BLOCK0):                Flash encryption mode counter                      = 127 R/- (0b1111111)
UART_DOWNLOAD_DIS (BLOCK0):              Disable UART download mode (ESP32 rev3 only)       = False R/- (0b0)
FLASH_CRYPT_CONFIG (BLOCK0):             Flash encryption config (key tweak bits)           = 15 R/W (0xf)
CONSOLE_DEBUG_DISABLE (BLOCK0):          Disable ROM BASIC interpreter fallback             = True R/W (0b1)
ABS_DONE_0 (BLOCK0):                     secure boot enabled for bootloader                 = False R/W (0b0)
ABS_DONE_1 (BLOCK0):                     secure boot abstract 1 locked                      = True R/W (0b1)
JTAG_DISABLE (BLOCK0):                   Disable JTAG                                       = True R/W (0b1)
DISABLE_DL_ENCRYPT (BLOCK0):             Disable flash encryption in UART bootloader        = True R/W (0b1)
DISABLE_DL_DECRYPT (BLOCK0):             Disable flash decryption in UART bootloader        = True R/W (0b1)
DISABLE_DL_CACHE (BLOCK0):               Disable flash cache in UART bootloader             = True R/W (0b1)
BLK1 (BLOCK1):                           Flash encryption key                              
   = ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? ?? -/- 
BLK2 (BLOCK2):                           Secure boot key                                   
   = the actual secure boot key was logged here R/- 
BLK3 (BLOCK3):                           Variable Block 3                                  
   = 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 R/W
```
**NOTE:** flashing the Secure Boot V1 hides the key in BLK2.

**Comment by projectgus** [comment]
Hi @achraf-boussaada ,

Glad that the workaround is working for you.

>     Does applying this workaround has any security implications?

Applying the workaround doesn't make the ESP32 vulnerable to any known issues. However starting from V4.3 we're going to recommend disabling UART Download Mode completely whenever Secure Boot V2 is enabled. Disabling UART download mode reduces the attack surface available for an attacker who is trying to bypass Secure Boot.

(For example: Earlier this year some additional fault injection issues were reported with Secure Boot V1  that relied on the UART download mode interface (although these attacks do not apply to Secure Boot V2, [see our advisory here](https://www.espressif.com/sites/default/files/advisory_downloads/Security%20Advisory%20CVE-2020-15048%2C%2013629%20EN%26CN.pdf).)

There are two possible ways to get the recommended configuration:

It is possible to re-enable this option later and OTA update to an app compiled with the option enabled. The app will disable UART ROM Download Mode on first boot, regardless of the bootloader config. Just be careful not to recompile the bootloader with this option enabled, unless you've already updated to an ESP-IDF version with the fix.

Alternatively the bug fix should be available on master branch quite soon, if you can afford to wait until it's available then this is the easiest and safest approach.

>  BLK2 contains now the Secure Boot key. The problem here is that the key itself is visible when calling espefuse.py summary and not replaced by question marks. Is it meant to be like that? Here's the eFuses table after flashing:

This is OK and intended.

In Secure Boot V1, the secure boot key is a secret key. Possession of this key allows generating valid Secure Boot digests for different bootloaders.

In Secure Boot V2, we moved to asymmetric encryption (public/private key pairs). The value stored in BLK2 is the hash of the public key used to verify the signature. Even possession of the entire public key (stored along with the signature in the bootloader binary) is not enough to sign a new bootloader or app, the corresponding *private key* is needed for this and it is not part of the firmware at all.

It's necessary for BLK2 to remain readable by software in this case because this is how the signature verification mechanism ensures that the provided public key is trusted (the SHA-256 hash of the public key bytes must match the hash stored here).

**Comment by monkeytronics** [comment]
Hi folks. Stumbled across this as I am seeing the same error I believe. Can I ask if this is fixed in 4.3.1?
